### PR TITLE
Inconsistent combat log sprites

### LINKED ISSUE #2186: Unit sprites in the Combat Log are inconsistent when multiple players connected.
Casual Joker is testing a SNAPSHOT of 47.10.  When he is hosting he can't see the sprites in game but the other players can.

@Krashner 

Also if he turns it off it turns it off for connected players.

Reference to PR https://github.com/MegaMek/megamek/pull/2135

### PR DESCRIPTION
**Pull request by Krashner**
This is a fix #2186. The problem was that the value for the option to add the sprites to the log was was checked when the image code was added to the reports instead of when the client requested the reports to fill the log. This resulted in them not being added for other clients when the option was disabled and it stopped them from showing up retroactively if reports were added while the option was disabled. Some null checks were added as well to prevent other potential issues.

This issue also mentioned that the sprites would not display at all for one user. This seems to be an issue in specific versions of Java. I was able to reproduce it when I installed a different version initially, but the issue seemed to go away after updating it. If this continues to be a problem, there's a potential workaround for displaying the sprites in the log. Instead of converting the images to Base64, the BufferedImages could instead be written to a temp folder and displayed by returning their file address when BASE64ImageView calls "getImageURL".

Also, this potentially includes a fix for #2190. I had encountered a StringIndexOutOfBoundsException on line 492 of the Report class, like in the issue. So I added a check to make sure it wasn't trying to read text out of bounds before inserting the code for the sprites.

### CODE REVIEW

File: megamek/src/megamek/common/Report.java [hunk 1]
```diff
@@ -489,7 +487,7 @@ public String getText() {
             }
             //draw the sprite at the beginning of the line
             if(imageCode != null && imageCode.length() > 0){
-                if(text.substring(0,1).equals("\n")){
+                if(text.length() >= 2 && text.substring(0,1).equals("\n")){
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/Report.java:490]: Is `text.startsWith("\n")` not an option? Simpler than a length check and first char check.

File: megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java [hunk 2]
```diff
@@ -51,25 +51,31 @@ private void populateImage() {
...
-        ByteArrayInputStream bais = null;
-        try {
-            bais = new ByteArrayInputStream(
-                    Base64.getDecoder().decode(b64.getBytes()));
-            newImage = ImageIO.read(bais);
-        } catch (Throwable ex) {
-            ex.printStackTrace();
+        if(b64 != null) {
+            BufferedImage newImage = null;
+            ByteArrayInputStream bais = null;
```
Comment (sixlettervariables) [inline megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java:66]: Are we leaking these BAIS? Or do we need them held open for the BufferedImage's.

Comment (Krashner) [reply megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java:66]: They should be closed, I'll fix that.

File: megamek/src/megamek/common/Report.java (hunk 1 above)
Comment (Krashner) [reply megamek/src/megamek/common/Report.java:490]: Ah, good call. That flew right past me.

File: megamek/src/megamek/client/Client.java [hunk 3]
```diff
@@ -1139,7 +1140,8 @@ public String receiveReport(Vector<Report> v) {
      * returns the stored <img> tag for given unit id
      */
     private String getCachedImgTag(int id){
-        if(imgCache == null || !imgCache.containsKey(id)) {
+        if(!GUIPreferences.getInstance().getBoolean(GUIPreferences.ADVANCED_ROUND_REPORT_SPRITES)
```
Comment (Windchild292) [inline megamek/src/megamek/client/Client.java:1143]:
```suggestion
        if (!GUIPreferences.getInstance().getBoolean(GUIPreferences.ADVANCED_ROUND_REPORT_SPRITES)
```

File: megamek/src/megamek/client/Client.java [hunk 4]
```diff
@@ -1139,7 +1140,8 @@ public String receiveReport(Vector<Report> v) {
...
+                || imgCache == null || !imgCache.containsKey(id)) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/Client.java:1144]:
```suggestion
                || (imgCache == null) || !imgCache.containsKey(id)) {
```

File: megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java [hunk 5]
```diff
@@ -51,25 +51,32 @@ private void populateImage() {
             cache = new Hashtable<>();
             getDocument().putProperty("imageCache", cache);
         }
-
         URL src = getImageURL();
-        cache.put(src, loadImage());
-
+        Image image = loadImage();
+        if(image != null) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java:56]:
```suggestion
        if (image != null) {
```

File: megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java [hunk 6]
```diff
@@ -51,25 +51,32 @@ private void populateImage() {
...
         String b64 = getBASE64Image();
-        BufferedImage newImage = null;
-        ByteArrayInputStream bais = null;
-        try {
-            bais = new ByteArrayInputStream(
-                    Base64.getDecoder().decode(b64.getBytes()));
-            newImage = ImageIO.read(bais);
-        } catch (Throwable ex) {
-            ex.printStackTrace();
+        if(b64 != null) {
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java:64]:
```suggestion
        if (b64 != null) {
```

File: megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java [hunk 7]
```diff
@@ -51,25 +51,32 @@ private void populateImage() {
...
+            BufferedImage newImage = null;
+            ByteArrayInputStream bais = null;
+            try {
+                bais = new ByteArrayInputStream(
+                        Base64.getDecoder().decode(b64.getBytes()));
+                newImage = ImageIO.read(bais);
+                bais.close();
+            } catch (Throwable ex) {
+                ex.printStackTrace();
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java:73]:
```suggestion
            } catch (Exception ex) {
                MegaMek.getLogger(getClass(), "loadImage", ex);
```

File: megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java [hunk 8]
```diff
@@ -51,25 +51,32 @@ private void populateImage() {
...
+            }
+            return newImage;
+        }
+        else{
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java:77]:
```suggestion
        } else {
```

File: megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java [hunk 9]
```diff
@@ -51,25 +51,32 @@ private void populateImage() {
...
+            return  null;
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java:78]:
```suggestion
            return null;
```

File: megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java [hunk 10]
```diff
@@ -51,25 +51,32 @@ private void populateImage() {
...
-        } catch (Throwable ex) {
-            ex.printStackTrace();
+        if(b64 != null) {
+            BufferedImage newImage = null;
+            ByteArrayInputStream bais = null;
+            try {
+                bais = new ByteArrayInputStream(
+                        Base64.getDecoder().decode(b64.getBytes()));
+                newImage = ImageIO.read(bais);
+                bais.close();
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java:71]: Can this be done through try with properties?

File: megamek/src/megamek/common/Report.java [hunk 11]
```diff
@@ -314,9 +313,8 @@ public void choose(boolean choice) {
      */
     public void addDesc(Entity entity) {
         if (entity != null) {
-            if(GUIPreferences.getInstance().getBoolean(GUIPreferences.ADVANCED_ROUND_REPORT_SPRITES)
-                    && (indentation <= Report.DEFAULT_INDENTATION || showImage)) {
-                imageCode = "<span id='" +entity.getId() + "'></span>";
+            if(indentation <= Report.DEFAULT_INDENTATION || showImage) {
```
Comment (Windchild292) [inline megamek/src/megamek/common/Report.java:316]:
```suggestion
            if ((indentation <= Report.DEFAULT_INDENTATION) || showImage) {
```

File: megamek/src/megamek/common/Report.java [hunk 12]
```diff
@@ -487,9 +485,9 @@ public String getText() {
                 }
                 i++;
             }
-            //draw the sprite at the beginning of the line
-            if(imageCode != null && imageCode.length() > 0){
-                if(text.substring(0,1).equals("\n")){
+            //add the sprite code at the beginning of the line
+            if(!imageCode.isEmpty()){
+                if(text.toString().startsWith("\n")){
```
Comment (Windchild292) [inline megamek/src/megamek/common/Report.java:490]:
```suggestion
            if (!imageCode.isEmpty()) {
                if (text.toString().startsWith("\n")) {
```

**Review by Windchild292 — COMMENTED** [review]
Bunch of formatting changes suggested (for future, please see the https://github.com/MegaMek/megamek/wiki/MegaMek-Coding-Style-Guide ).

Also a few suggestions and one code improvement idea.

File: megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java (hunk 10 above)
Comment (Krashner) [reply megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java]: That's a good call, I'll write it in.

File: megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java [hunk 13]
```diff
@@ -53,29 +54,25 @@ private void populateImage() {
...
-                bais = new ByteArrayInputStream(
-                        Base64.getDecoder().decode(b64.getBytes()));
+            try (ByteArrayInputStream bais = new ByteArrayInputStream(
+                        Base64.getDecoder().decode(b64.getBytes()))) {
                 newImage = ImageIO.read(bais);
-                bais.close();
-            } catch (Throwable ex) {
-                ex.printStackTrace();
+            } catch (Exception ex) {
+                MegaMek.getLogger().error(this.getClass(), "loadImage()", ex);
```
Comment (Windchild292) [inline megamek/src/megamek/client/ui/swing/util/BASE64ImageView.java:71]:
```suggestion
                MegaMek.getLogger().error(this, ex);
```

**Review by Windchild292 — APPROVED** [review]
Approved pending CI

**Comment by HammerGS** [comment]
Not sure if it's to late, but this is going to be a very popular option based on the feedback and teases.

Could we move the the option off the Advanced tab to the Main Tab perhaps under the Second section Tooltips or Graphics some where.

**Comment by NickAragua** [comment]
Let's make that a separate pull request. I personally think it'll be fine in the advanced settings, just make it on by default.

**Comment by HammerGS** [comment]
> Let's make that a separate pull request. I personally think it'll be fine in the advanced settings, just make it on by default.

That works for me.

**Pull request merged**